Applications/Spire/Source/Ui/TagBox.cpp

```diff
@@ -0,0 +1,480 @@
+    : m_source(std::move(source)) {}
+
+  virtual int get_size() const {
+    return m_source->get_size() + 2;
+  }
+
+  virtual const std::any& at(int index) const {
+    if(index < 0 || index >= get_size()) {
+      throw std::out_of_range("The index is out of range.");
+    }
+    if(index < m_source->get_size()) {
+      return m_source->at(index);
+    }
+    static auto value = std::any();
+    return value;
+  }
+
+  virtual QValidator::State set(int index, const std::any& value) {
+    if(index < 0 || index >= get_size()) {
+      return QValidator::State::Invalid;
+    }
+    if(index < m_source->get_size()) {
+      return m_source->set(index, value);
+    }
+    return QValidator::State::Acceptable;
+  }
+
+  connection connect_operation_signal(
+    const OperationSignal::slot_type& slot) const override {
+    return m_source->connect_operation_signal(slot);
+  }
+};
+
+TagBox::TagBox(std::shared_ptr<ListModel> list_model,
+    std::shared_ptr<TextModel> current_model, QWidget* parent)
+    : QWidget(parent),
+      m_model(std::make_shared<PartialListModel>(std::move(list_model))),
+      m_focus_observer(*this),
+      m_overflow(TagBoxOverflow::WRAP),
+      m_tags_width(0),
+      m_list_item_gap(0) {
+  auto layout = new QHBoxLayout(this);
+  layout->setContentsMargins({});
+  m_text_box = new TextBox(std::move(current_model));
+  set_style(*m_text_box, TEXT_BOX_STYLE(get_style(*m_text_box)));
+  m_text_box->installEventFilter(this);
+  m_list_view = new ListView(m_model,
+    std::bind_front(&TagBox::build_tag, this));
+  m_list_view->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
+  set_style(*m_list_view, LIST_VIEW_STYLE(get_style(*m_list_view)));
+  m_list_view_style_connection = connect_style_signal(*m_list_view,
+    std::bind_front(&TagBox::on_list_view_style, this));
+  for(auto i = 0; i < m_list_view->get_list()->get_size(); ++i) {
+    set_style(*m_list_view->get_list_item(i), LIST_ITEM_STYLE());
+  }
+  m_list_view->get_list()->connect_operation_signal(
+    std::bind_front(&TagBox::on_operation, this));
+  m_list_view->connect_submit_signal(std::bind_front(&TagBox::on_submit, this));
+  m_list_view->setFocusPolicy(Qt::NoFocus);
+  m_list_view->installEventFilter(this);
+  m_list_view_container = new QWidget();
+  auto container_layout = new QHBoxLayout(m_list_view_container);
+  container_layout->setContentsMargins({});
+  container_layout->setSpacing(0);
+  container_layout->addWidget(m_list_view);
+  auto input_box = make_input_box(m_list_view_container);
+  layout->addWidget(input_box);
+  proxy_style(*this, *input_box);
+  set_style(*this, INPUT_BOX_STYLE(get_style(*input_box)));
+  m_style_connection = connect_style_signal(*this, [=] { on_style(); });
+  m_ellipses_item = m_list_view->get_list_item(get_list()->get_size());
+  m_ellipses_item->hide();
+  setFocusProxy(m_text_box);
+  setFocusPolicy(Qt::StrongFocus);
+  m_line_edit = m_text_box->findChild<QLineEdit*>();
```

Comment:
Remove the use of ```findChild<QLineEdit*>```, instead use ```TextBox::get_selection()```.

Reply:
I have merged the latest commit, but I don't find TextBox::get_selection().